### PR TITLE
Add documentation for informal Ruby memory model and synchronisation …

### PR DESCRIPTION
**Pull request by pitr-ch**
…extensions.

### CODE REVIEW

**Comment by pitr-ch** [comment]
Hi I am sharing the Ruby informal memory model for first round of comments, then I intend to widen the audience after first round of comments is addressed and silly mistakes removed. Thanks for taking time to look at it. Please invite whoever you would like to, just do not tweet (etc.) yet.

cc @chrisseaton @eregon @jdantonio @headius.

File: doc/synchronization.md [hunk 1]
```diff
@@ -0,0 +1,356 @@
+> Quotations are used for notes.
+
+> This document is work-in-progress.
+> Intentions of this effort and document are: to sum up the behavior
```
Comment (chrisseaton) [inline doc/synchronization.md]: Could be 'to summarise' rather than 'to sum up'. Sum up sounds more like a conclusion.

File: doc/synchronization.md [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+> of Ruby in concurrent and parallel environment, initiate discussion,
+> identify problems in the document, find flaws in the Ruby
+> implementations if any, suggest what has to be enhanced in Ruby itself
+> and cooperate towards the goal in all implementations (using
+> `concurrent-ruby` as compatibility layer).
```
Comment (chrisseaton) [inline doc/synchronization.md]: Maybe as in a JEP, list non-goals as well as goals to limit scope creep. For example it is a non-goal to add high-level new language features.

File: doc/synchronization.md [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+# Synchronization
+
+This layer provides tools to write concurrent abstractions independent on any
+particular Ruby implementation. It is build on top of Ruby memory model which is
+also described here. `concurrent-ruby` abstractions are build using this layer.
+
+**Why?** Ruby is great expressive language, but it lacks in support for
```
Comment (chrisseaton) [inline doc/synchronization.md]: I think it 'lacks in support for well-defined low-level concurrent and parallel computation', rather than any parallel or concurrent computation, as it can do that with locks etc.

File: doc/synchronization.md [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+concurrent and parallel computation. It's hoped that this document will provide
+ground steps for Ruby to become as good in this area as in others.
+
+Without memory model and this layer it's very hard to write concurrent
```
Comment (chrisseaton) [inline doc/synchronization.md]: 'Without *a* memory model'

File: doc/synchronization.md [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+abstractions for Ruby. To write a proper concurrent abstraction it often means
+to reimplement it more than once for different Ruby runtimes, which is very
+time-consuming and error-prone.
+
+# Ruby memory model
+
+Ruby memory model is a framework allowing to reason about programs in concurrent
```
Comment (chrisseaton) [inline doc/synchronization.md]: '*The* Ruby memory model'

File: doc/synchronization.md [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+and parallel environment. It allows to identify what is and what is not a [race
+condition](https://en.wikipedia.org/wiki/Race_condition). Memory model is also a
+contract: if a program is written without the race conditions it behaves in
+[sequential consistent](https://en.wikipedia.org/wiki/Sequential_consistency)
+manner.
```
Comment (chrisseaton) [inline doc/synchronization.md]: In general 'memory model' always needs an article - either 'the' memory model, or 'a' memory model, but never simply 'memory model'.

File: doc/synchronization.md [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+abstractions for Ruby. To write a proper concurrent abstraction it often means
+to reimplement it more than once for different Ruby runtimes, which is very
+time-consuming and error-prone.
+
+# Ruby memory model
+
+Ruby memory model is a framework allowing to reason about programs in concurrent
+and parallel environment. It allows to identify what is and what is not a [race
+condition](https://en.wikipedia.org/wiki/Race_condition). Memory model is also a
+contract: if a program is written without the race conditions it behaves in
```
Comment (chrisseaton) [inline doc/synchronization.md]: But 'race conditions' does not need an article as it's plural :) So simply 'without race conditions'

Comment (chrisseaton) [inline doc/synchronization.md]: I'm getting a bit pedantic here, but race conditions are not simply an evil thing. There are several types of race conditions, and here we probably mean data races. Even data races can be part of a well designed application. For example in Go if you have two go-routines and select on them both to get a single value, then that's a race, but that may be what you want.

From the Encylopedia of Parallel Computing (David Padua ed.):

> A race condition occurs in a parallel program execution when two or more threads access a common resource, e.g., a variable in shared memory, and the order of the accesses depends on the timing, i.e., the progress of individual threads.
> The disposition for a race condition is in the parallel program. In different executions of the same program on the same input access events that constitute a race can occur in different order, which may but does not generally result in different program behaviors (non- determinacy).
> Race conditions that can cause unintended non- determinacy are programming errors. Examples of such programming errors are violations of atomicity due to incorrect synchronization.

So a race condition is not an inherently error! It's just a property of the program.

File: doc/synchronization.md [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+> Intentions of this effort and document are: to sum up the behavior
+> of Ruby in concurrent and parallel environment, initiate discussion,
+> identify problems in the document, find flaws in the Ruby
+> implementations if any, suggest what has to be enhanced in Ruby itself
+> and cooperate towards the goal in all implementations (using
+> `concurrent-ruby` as compatibility layer).
+
+# Synchronization
+
+This layer provides tools to write concurrent abstractions independent on any
```
Comment (eregon) [inline doc/synchronization.md]: on => of

File: doc/synchronization.md [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+manner.
+
+Sources:
+
+-   [Java memory model](http://www.cs.umd.edu/~pugh/java/memoryModel/),
+    and its [FAQ](http://www.cs.umd.edu/~pugh/java/memoryModel/jsr-133-faq.html)
+-   [atomic&lt;&gt; Weapons 1](https://channel9.msdn.com/Shows/Going+Deep/Cpp-and-Beyond-2012-Herb-Sutter-atomic-Weapons-1-of-2)
+and
+[2](https://channel9.msdn.com/Shows/Going+Deep/Cpp-and-Beyond-2012-Herb-Sutter-atomic-Weapons-2-of-2)
+- [JRuby's wiki page about concurrency](https://github.com/jruby/jruby/wiki/Concurrency-in-jruby)
```
Comment (chrisseaton) [inline doc/synchronization.md]: I think we should think about how we talk about JRuby, Java and the JMM. A concern from casual readers could be that we are putting the JMM into Ruby or forcing the JRuby way of doing things onto MRI. How do we handle this question?

File: doc/synchronization.md [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+> of Ruby in concurrent and parallel environment, initiate discussion,
+> identify problems in the document, find flaws in the Ruby
+> implementations if any, suggest what has to be enhanced in Ruby itself
+> and cooperate towards the goal in all implementations (using
+> `concurrent-ruby` as compatibility layer).
+
+# Synchronization
+
+This layer provides tools to write concurrent abstractions independent on any
+particular Ruby implementation. It is build on top of Ruby memory model which is
```
Comment (eregon) [inline doc/synchronization.md]: built ... the Ruby MM ...

File: doc/synchronization.md [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+[MRI](https://www.ruby-lang.org/en/), [JRuby](http://jruby.org/),
+[JRuby+Truffle](https://github.com/jruby/jruby/wiki/Truffle),
+[Rubinius](http://rubini.us/); taking account limitations of the implementations
+or their platforms; inspiration drawn from other existing memory models (Java,
+C++11). This is not a formal model.
+
+Key properties are:
+
+-   **volatility (V)** - Any written value is immediately visible to any
+    subsequent volatile reads including all writes leading to this value. (Same
```
Comment (chrisseaton) [inline doc/synchronization.md]: Not sure what 'including all writes leading to this value' means

File: doc/synchronization.md [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+    writes, implies an update cannot be lost).
+-   Module/Class definition is actually constant definition. It is defined
+    instantly, however its methods are then processed sequentially.
+-   `||=`, `+=`, etc. are actually two operations read and write which implies
+    that it's not an atomic operation. See volatile variables
+    with compare-and-set.
+-   Method invocation does not have any special properties that includes
+    object initialization.
+
+Implementation differences from the model:
```
Comment (chrisseaton) [inline doc/synchronization.md]: Differences at the moment, or permanent differences?

File: doc/synchronization.md [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+| include/extend | x | x | x | If `AClass` is included `AModule`, `AClass` gets all `AModule`'s methods at once. |
+
+
+Notes:
+
+-   Variable read reads value from preexisting variable.
+-   Variable definition creates new variable (operation is serialized with
+    writes, implies an update cannot be lost).
+-   Module/Class definition is actually constant definition. It is defined
+    instantly, however its methods are then processed sequentially.
```
Comment (chrisseaton) [inline doc/synchronization.md]: I don't understand the difference between 'instantly' and 'sequentially'.

File: doc/synchronization.md [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+    with compare-and-set.
+-   Method invocation does not have any special properties that includes
+    object initialization.
+
+Implementation differences from the model:
+
+-   MRI: everything is volatile.
+-   JRuby: `Thread` and `Fiber` local variables are volatile. Instance
+    variables are volatile on x86 and people may un/intentionally depend
+    on the fact.
```
Comment (chrisseaton) [inline doc/synchronization.md]: Why volatile on x86 but not other platforms?

File: doc/synchronization.md [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+
+-   Beware of requiring and autoloading in concurrent programs, it's possible to
+    see partially defined classes. Eager loading or blocking until class is
+    fully loaded has should be used to mitigate.
+
+### Core classes
+
+`Mutex`, `Monitor`, `Queue` have to work correctly on each implementation. Ruby
+implementations should not crash when e.g. Array is used in parallel environment
+but it may loose updates etc.
```
Comment (chrisseaton) [inline doc/synchronization.md]: I think this is absolutely the right decision - but I think it needs some narrative to justify. We do not syncohronise Array etc on your behalf. You need to use locks on them when you use them concurrently, but this means you do not.

Also, what does crash mean? JRuby will give you concurrent modification exceptions, which are normal Ruby exceptions. I think Rubinius might as well. Does that count as a crash?

File: doc/synchronization.md [hunk 16]
```diff
@@ -0,0 +1,356 @@
...
+Key properties are:
+
+-   **volatility (V)** - Any written value is immediately visible to any
+    subsequent volatile reads including all writes leading to this value. (Same
+    meaning as in Java.)
+-   **atomicity (A)** - Operation is either done or not as a whole.
+-   **serialized (S)** - Operations are serialized in some order (they
+    cannot disappear). This is a new property not mentioned in other memory
+    models, since Java and C++ do not have dynamically defined fields. All
+    operations on one line in the table are serialized with each other.
```
Comment (eregon) [inline doc/synchronization.md]: in a row

File: doc/synchronization.md [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+class ImmutableTreeNode < Concurrent::Synchronization::Object
+  # mark this class to publish final instance variables safely
+  safe_initialization!
+
+  def initialize(left, right)
+    # Call super to allow proper initialization.
+    super()
+    # By convention final instance variables have CamelCase names
+    # to distinguish them from ordinary instance variables.
+    @Left  = left
```
Comment (chrisseaton) [inline doc/synchronization.md]: Does this capitalisation mean something?

File: doc/synchronization.md [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+
+## Volatile instance variable
+
+`Synchronization::Object` children can have volatile instance variables. A Ruby
+library cannot alter meaning of `@a_name` expression therefore when a
+`attr_volatile :a_name` is called, declaring the instance variable named
+`a_name` to be volatile, it creates method accessors.
+
+> However there is Ruby [issue](https://redmine.ruby-lang.org/issues/11539)
+> filled to address this.
```
Comment (chrisseaton) [inline doc/synchronization.md]: '*filed*'

File: doc/synchronization.md [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+-   **serialized (S)** - Operations are serialized in some order (they
+    cannot disappear). This is a new property not mentioned in other memory
+    models, since Java and C++ do not have dynamically defined fields. All
+    operations on one line in the table are serialized with each other.
+
+### Core behavior:
+
+| Operation | V | A | S | Notes |
+|:----------|:-:|:-:|:-:|:-----|
+| local variable read/write/definition | - | x | x | Local variables are determined during parsing, they are not usually dinamically added (with exception of `local_variable_set`). Therefore definition is quite rare. |
```
Comment (eregon) [inline doc/synchronization.md]: dynamically.

File: doc/synchronization.md [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+| instance variable read/write/(un)definition | - | x | x ||
+| class variable read/write/(un)definition | x | x | x ||
+| global variable read/write/definition | x | x | x | un-define us not possible currently. |
+| constant variable read/write/(un)definition | x | x | x ||
+| `Thread` local variable read/write/definition | - | x | x | un-define us not possible currently. |
```
Comment (eregon) [inline doc/synchronization.md]: *is* not possible

File: doc/synchronization.md [hunk 21]
```diff
@@ -0,0 +1,356 @@
...
+    models, since Java and C++ do not have dynamically defined fields. All
+    operations on one line in the table are serialized with each other.
+
+### Core behavior:
+
+| Operation | V | A | S | Notes |
+|:----------|:-:|:-:|:-:|:-----|
+| local variable read/write/definition | - | x | x | Local variables are determined during parsing, they are not usually dinamically added (with exception of `local_variable_set`). Therefore definition is quite rare. |
+| instance variable read/write/(un)definition | - | x | x ||
+| class variable read/write/(un)definition | x | x | x ||
```
Comment (eregon) [inline doc/synchronization.md]: Why this is magically volatile? class vars are also mostly deprecated AFAIK.

File: doc/synchronization.md [hunk 22]
```diff
@@ -0,0 +1,356 @@
...
+| `Fiber` local variable read/write/definition | - | x | x | un-define us not possible currently. |
+| method creation/redefinition/removal | x | x | x ||
+| include/extend | x | x | x | If `AClass` is included `AModule`, `AClass` gets all `AModule`'s methods at once. |
+
+
+Notes:
+
+-   Variable read reads value from preexisting variable.
+-   Variable definition creates new variable (operation is serialized with
+    writes, implies an update cannot be lost).
```
Comment (eregon) [inline doc/synchronization.md]: For all variables? even locals? or just ivars?

File: doc/synchronization.md [hunk 23]
```diff
@@ -0,0 +1,356 @@
...
+    cannot disappear). This is a new property not mentioned in other memory
+    models, since Java and C++ do not have dynamically defined fields. All
+    operations on one line in the table are serialized with each other.
+
+### Core behavior:
+
+| Operation | V | A | S | Notes |
+|:----------|:-:|:-:|:-:|:-----|
+| local variable read/write/definition | - | x | x | Local variables are determined during parsing, they are not usually dinamically added (with exception of `local_variable_set`). Therefore definition is quite rare. |
+| instance variable read/write/(un)definition | - | x | x ||
```
Comment (eregon) [inline doc/synchronization.md]: Maybe this should mention that an ivar which is defined must appear eventually?

File: doc/synchronization.md (hunk 14 above)
Comment (eregon) [inline doc/synchronization.md]: The fact that you can use `Thread#[]` or `Thread#thread_variable_get` on another thread seems like bad design. Indeed, not guaranteeing that it is volatile makes sense to me.

File: doc/synchronization.md [hunk 24]
```diff
@@ -0,0 +1,356 @@
...
+    variables are volatile on x86 and people may un/intentionally depend
+    on the fact.
+-   Class variables require investigation.
+
+### Source loading:
+
+| Operation | V | A | S | Notes |
+|:----------|:-:|:-:|:-:|:-----|
+| requiring | x | x | x | File will not be required twice, classes and modules are still defined gradually. |
+| autoload | x | x | x | Only one autoload at a time. |
```
Comment (eregon) [inline doc/synchronization.md]: Why only one at a time? To ensure partial classes are not seen?

File: doc/synchronization.md [hunk 25]
```diff
@@ -0,0 +1,356 @@
...
+
+Notes:
+
+-   Beware of requiring and autoloading in concurrent programs, it's possible to
+    see partially defined classes. Eager loading or blocking until class is
+    fully loaded has should be used to mitigate.
```
Comment (eregon) [inline doc/synchronization.md]: -has

File: doc/synchronization.md [hunk 26]
```diff
@@ -0,0 +1,356 @@
...
+    self.value = value
+  end
+
+  # Very cheap reader of the Counter's current value, just a volatile read.
+  def count
+    # volatile read
+    value
+  end
+
+  # Safely increment the value without loosing updates
```
Comment (eregon) [inline doc/synchronization.md]: increments

File: doc/synchronization.md [hunk 27]
```diff
@@ -0,0 +1,356 @@
...
+  # (as it would happen with just += used).
+  def increment(add)
+    # Wrap he two volatile operations to make them atomic.
```
Comment (eregon) [inline doc/synchronization.md]: Wrap the

File: doc/synchronization.md [hunk 28]
```diff
@@ -0,0 +1,356 @@
...
+      # volatile write and read
+      self.value = self.value + add
+    end
+  end   
+end
+```
+
+> This is currently planned to be migrated to a module to be able to add
+> volatile fields any object not just `Synchronization::Object` children. The
+> instance variable itself is named `@"volatile_#(name)"` to distinguish it and
```
Comment (eregon) [inline doc/synchronization.md]: interesting notation, I would go with either the quotes outside and #{} or just `@volatile_<name>`

File: doc/synchronization.md [hunk 29]
```diff
@@ -0,0 +1,356 @@
...
+## Concurrent Ruby Notes
+
+### Locks
+
+Concurrent Ruby also has an internal extension of `Object` called
+`LockableObject`, which provides same synchronization primitives as Java's
+Object: `synchronize(&block)`, `wait(timeout = nil)`,
+`wait_until(timeout = nil, &condition)`, `signal`, `broadcast`. This class is
+intended for internal use in `concurrent-ruby` only and it does not support
+subclassing (since it cannot protect its lock from its children, for more
```
Comment (eregon) [inline doc/synchronization.md]: Is this checked? via a inherited/included hook throwing an exception?

File: doc/synchronization.md [hunk 30]
```diff
@@ -0,0 +1,356 @@
...
+to allow to call methods in the same object without double locking.
+
+``` ruby
+class Node
+  # ...
+  def left
+    synchronize { ns_left }
+  end  
+
+  def left
```
Comment (eregon) [inline doc/synchronization.md]: right + below

File: doc/synchronization.md [hunk 31]
```diff
@@ -0,0 +1,356 @@
...
+```
+
+Three of them were used in the example above.
+
+> Current implementation relies on final instance variables where a instance of
+> `AtomicReference` is held to provide compare-and-set operations. That creates
+> extra indirection which is hoped to be removed over time when better
+> implementation will become available in Ruby implementations.
+
+## Concurrent Ruby Notes
```
Comment (eregon) [inline doc/synchronization.md]: This should be another document

File: doc/synchronization.md (hunk 14 above)
Comment (eregon) [inline doc/synchronization.md]: Also it is not proved in the current JRuby implementation, is it?

File: doc/synchronization.md (hunk 7 above)
Comment (jdantonio) [inline doc/synchronization.md]: :+1: I don't think that's pedantic at all. The Ruby community, in general, has a very poor understanding of concurrency. We should be very specific with our verbiage so that we don't create additional confusion.

File: doc/synchronization.md (hunk 9 above)
Comment (jdantonio) [inline doc/synchronization.md]: :+1: That's why the README and other docs generally use words like "inspire" and "influence."

File: doc/synchronization.md (hunk 28 above)
Comment (jdantonio) [inline doc/synchronization.md]: I don't think the quotes are necessary at all, the backticks provide enough highlighting.

File: doc/synchronization.md (hunk 11 above)
Comment (pitr-ch) [inline doc/synchronization.md]: I am pointing to a piggybacking technique, this explains it better: http://stackoverflow.com/questions/8769570/volatile-piggyback-is-this-enough-for-visiblity.

File: doc/synchronization.md (hunk 13 above)
Comment (pitr-ch) [inline doc/synchronization.md]: Bad wording. Is "Module or Class is defined atomically, then its methods are defined atomically one by one." better?

File: doc/synchronization.md (hunk 14 above)
Comment (pitr-ch) [inline doc/synchronization.md]: `Thread#[]` or `Thread#thread_variable_get` are synchronised methods on JRuby.

IVars on JRuby are not fully correct at the moment, there is loadFence missing (considering only path using fences), but @headius created an issue to fix it AFAIK. So for this document I've considered them already volatile. The volatile ivar parts was agreed (best to my understanding) to be removed from https://github.com/jruby/jruby/wiki/Concurrency-in-jruby to avoid users getting used to it.

File: doc/synchronization.md (hunk 24 above)
Comment (pitr-ch) [inline doc/synchronization.md]: Hm right you could be loading concurrently different constants, given constant can auto-loaded only once though, rest is blocking until the particular auto-load finishes. If you agree, I'll update this part.

File: doc/synchronization.md (hunk 28 above)
Comment (pitr-ch) [inline doc/synchronization.md]: Interesting indeed, I would say totally messed up :)

File: doc/synchronization.md (hunk 31 above)
Comment (pitr-ch) [inline doc/synchronization.md]: True, splitting.

File: doc/synchronization.md (hunk 29 above)
Comment (pitr-ch) [inline doc/synchronization.md]: Its a TODO to do that.

**Comment by pitr-ch** [comment]
Thanks for your comments, following commit will include only typo fixes, I'll comment on the more important questions after that.

File: doc/synchronization.md (hunk 2 above)
Comment (pitr-ch) [inline doc/synchronization.md]: Good idea, added.

File: doc/synchronization.md (hunk 9 above)
Comment (pitr-ch) [inline doc/synchronization.md]: I'll include explanation why there is a JRuby resource but not a one for MRI and Rubinius. Basically I could not find similar comprehensive page for MRI and Rubinius.

Comment (pitr-ch) [inline doc/synchronization.md]: Updated in d4dc311.

File: doc/synchronization.md (hunk 7 above)
Comment (pitr-ch) [inline doc/synchronization.md]: Yeah I did not get this part quite right, improved in f0c4cce.

File: doc/synchronization.md (hunk 11 above)
Comment (pitr-ch) [inline doc/synchronization.md]: Updated in b2ba23a.

File: doc/synchronization.md (hunk 15 above)
Comment (pitr-ch) [inline doc/synchronization.md]: Updated in 8af0aa6.

File: doc/synchronization.md (hunk 21 above)
Comment (pitr-ch) [inline doc/synchronization.md]: There were 2 things I wanted to discuss more, but you've already asked about both :) First is class variables. Since they are mostly deprecated I did not spend much time investigation, what the current behaviour is. AFAIK: MRI: volatile, JRuby: volatile, rbx: ?, J+T: non-volatile.

I am considering following:

- they are deprecated, not that important to be performant. -> volatile
- they are shared across child classes making them volatile could be quite expensive -> non-volatile
- if non-volatile, extension creating volatile method accessors can be added. -> non-volatile
- to best my knowledge they are usually used for a shared global flag where volatility is required. -> volatile
- there is no extension to make it volatile. -> volatile

I think the correct solution would be having them non-volatile. On the other hand it may not be worth it since they used rarely and deprecated. I would like to hear your opinions.

File: doc/synchronization.md (hunk 22 above)
Comment (pitr-ch) [inline doc/synchronization.md]: This line is intended as a general explanation what is meant by definition of a variable and what does it imply when combined with variable write.

Comment (pitr-ch) [inline doc/synchronization.md]: Second thing I wanted to discuss more is: should local variables be serialised?

Local variables are determined during parsing, only problematic thing is `local_variable_set`. I thought I understand it but this example surprised me:
```ruby
[8] pry(main)> b=-> {a}
=> #<Proc:0x007fc68a76c430@(pry):8 (lambda)>
[9] pry(main)> b.binding.local_variable_set :a, 1
=> 1
[10] pry(main)> b.call
NameError: undefined local variable or method `a' for main:Object
```
This works though:
```
[11] pry(main)> a = 0
=> 0
[12] pry(main)> b=-> {a}
=> #<Proc:0x007fc68e922d70@(pry):12 (lambda)>
[13] pry(main)> b.binding.local_variable_set :a, 1
=> 1
[14] pry(main)> b.call
=> 1
```
So maybe a local variable cannot be really defined? More information here http://stackoverflow.com/questions/19321698/in-ruby-is-there-no-way-to-dynamically-define-a-local-variable-in-the-current-c .

Comment (chrisseaton) [inline doc/synchronization.md]: When line 8 is parsed, `a` isn't a local variable, so it's assumed to be a method call. The error message says 'undefined local variable or method `a'', but really I think the decision has already been taken that it is a method. When you add a local variable, it doesn't go back and reconsider what `a` could be.

File: doc/synchronization.md [hunk 32]
```diff
@@ -0,0 +1,340 @@
...
+Key properties are:
+
+-   **volatility (V)** - Any written value is immediately visible to any
+    subsequent volatile read of the same variable. Any write executed before
+    volatile write based on program-order is visible to the read as well, which
+    allows
+    [piggybacking](http://stackoverflow.com/questions/8769570/volatile-piggyback-is-this-enough-for-visiblity).
+    (Same meaning as in Java, it creates synchronizes-with (JMM term) order
+    between write and read, which participates in creating happens-before
+    order.)
```
Comment (eregon) [inline doc/synchronization.md]: This is very complex now, I think piggybacking is also going too far from the definition.
Maybe we could just say that 'subsequent read' is either same thread, later in the program, or there is some synchronization between the write and the subsequent read.

File: doc/synchronization.md (hunk 22 above)
Comment (eregon) [inline doc/synchronization.md]: One way to understand this is the binding has an additional frame of its own, so new variables can only be read from that binding. If the variable already exists in scope, it is however just set to the new value.

File: doc/synchronization.md (hunk 21 above)
Comment (eregon) [inline doc/synchronization.md]: How is it implemented in JRuby? Similar to ivars?
It probably make sense to have them volatile as they are some sort of "namespaced" global state.

Comment (pitr-ch) [inline doc/synchronization.md]: Currently the JRuby implementation is volatile but it breaks the serialization, see https://github.com/jruby/jruby/issues/3357

File: doc/synchronization.md (hunk 32 above)
Comment (pitr-ch) [inline doc/synchronization.md]: updated in 2b6a7bd

File: doc/synchronization.md (hunk 24 above)
Comment (pitr-ch) [inline doc/synchronization.md]: Updated 661a5b6.

**Comment by pitr-ch** [comment]
Thanks for your comments so far. It's getting quite complicated here to follow and to comment, therefore I'll move further discussion to Google docs. Two main remaining topics to solve is volatile class variables and serialised local variables.

**Comment by pitr-ch** [comment]
The document was migrated to https://docs.google.com/document/d/1pVzU8w_QF44YzUCCab990Q_WZOdhpKolCIHaiXG-sPw/edit?usp=sharing Please use it instead, I hope it'll be more pleasant to comment on.
